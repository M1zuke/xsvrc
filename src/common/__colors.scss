// Main Colors
$color0: #1A1A1A;
$color1: #303030;
$color2: #212121;
$color3: #666666;

$colorBlue: #25AAC5;
$colorWhite: #FFFFFF;
$textColor: #E7E7E7;
$colorBlack: #000000;
$colorGreen: #00FD00;
$colorRed: red;

// Index
$indexBackgroundColor: $color0;
$indexFontColor: $textColor;

// Login
$loginBackgroundColor: $color1;

// Loading
$loadingBackgroundColor: $textColor;

// Button
$buttonBackgroundColor: $color0;
$buttonFontColor: $textColor;
$buttonBackgroundColorOnHover: $color2;
$buttonActiveBackgroundColor: $colorBlue;
$buttonActiveFontColor: $colorBlack;
$buttonActiveBackgroundColorOnHover: $colorBlue;

// TextInput
$textInputBackgroundColor: $color2;
$textInputFontColor: $textColor;

// Navigation
$userOverviewBackgroundColor: $color1;
$userOverviewHighlight: $color2;
