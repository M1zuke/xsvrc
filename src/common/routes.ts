export type RouteDefinition = {
  config: {
    path: string;
    exact?: boolean;
  };
  path: string;
};

export type Routes = {
  home: RouteDefinition;
  login: RouteDefinition;
  friends: RouteDefinition;
  eventList: RouteDefinition;
  notifications: RouteDefinition;
  settings: RouteDefinition;
  moderation: RouteDefinition;
  avatar: RouteDefinition;
  userSearch: RouteDefinition;
  transactions: RouteDefinition;
};

export const routes: Routes = {
  home: { path: '/', config: { path: '/', exact: true } },
  login: { path: '/login', config: { path: '/login', exact: true } },
  friends: { path: '/friends', config: { path: '/friends', exact: true } },
  eventList: { path: '/user-event-list', config: { path: '/user-event-list', exact: true } },
  notifications: { path: '/notifications', config: { path: '/notifications', exact: true } },
  settings: { path: '/settings', config: { path: '/settings', exact: true } },
  moderation: { path: '/moderation', config: { path: '/moderation', exact: true } },
  avatar: { path: '/avatar', config: { path: '/avatar', exact: true } },
  userSearch: { path: '/user-search', config: { path: '/user-search', exact: true } },
  transactions: { path: '/transactions', config: { path: '/transactions', exact: true } },
};
