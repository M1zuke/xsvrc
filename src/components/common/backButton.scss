@import "../../utils/colors";

.back-btn-component {
  grid-area: BackBtn;
  background-color: $highlightColor;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  user-select: none;
  padding: 0 15px;
  border-radius: 5px;

  &:hover {
    background-color: $buttonHighlightColorHover;
  }
}
