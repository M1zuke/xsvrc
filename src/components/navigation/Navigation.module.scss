@import '../../common/_spacing';
@import '../../common/_mixins';
@import '../../common/_colors';

.Component {
  display: grid;
  margin: $innerSpacing 0 $innerSpacing $innerSpacing;
  grid-template-rows: auto autoSize();

  .Navigation {
    display: grid;
    grid-template-rows: repeat(8, autoSize());
    grid-gap: $tinySpacing;
  }

  .Settings {
    display: grid;
    grid-template-rows: repeat(2, autoSize());
    grid-gap: $tinySpacing;

    .LogoutButton {
      color: red;
    }
  }
}
