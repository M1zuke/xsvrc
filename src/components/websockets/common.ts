import { isLoaded } from '../../api/prepare';
import { UserInfo } from '../../api/types';
import { AppState } from '../../store';

export type Comparison<V> = {
  from: V;
  to: V;
};

export type UserComparison<K extends keyof UserInfo = keyof UserInfo> = Partial<{
  [P in K]: Comparison<UserInfo[P]>;
}>;

export function compareUsers(oldUser: UserInfo | null = null, newUser: UserInfo): UserComparison {
  if (oldUser === null) {
    return {};
  }
  return Object.assign(
    {},
    ...Object.keys(newUser).map((key) => {
      const k = key as keyof UserInfo;
      const v = oldUser[k];
      const nv = newUser[k];
      if (Array.isArray(v) && Array.isArray(nv)) {
        const arrayComparison = compareArray(v, nv);
        if (arrayComparison.added.length === 0 && arrayComparison.removed.length === 0) {
          return undefined;
        }
        return {
          [key]: {
            from: v,
            to: nv,
          },
        };
      }

      if (oldUser[k] !== undefined && oldUser[k] !== newUser[k]) {
        return {
          [key]: {
            from: oldUser[k],
            to: newUser[k],
          },
        };
      }
      return undefined;
    }),
  );
}

type ArrayComparison<T extends string> = {
  added: T[];
  removed: T[];
};

function compareArray<T extends string = string>(a: T[], b: T[]): ArrayComparison<T> {
  const differencesToA: T[] = [];
  b.forEach((v) => {
    if (!a.includes(v)) {
      differencesToA.push(v);
    }
  });

  const differencesToB: T[] = [];
  a.forEach((v) => {
    if (!b.includes(v)) {
      differencesToA.push(v);
    }
  });
  return { added: differencesToA, removed: differencesToB };
}

export function getFriendsAndOldUser(state: AppState, userId: string): [UserInfo[], UserInfo | null] {
  if (isLoaded(state.friends.friendInfo)) {
    return [Object.values(state.friends.friendInfo), state.friends.friendInfo[userId] || null];
  }
  return [[], null];
}
