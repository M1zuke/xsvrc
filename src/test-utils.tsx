import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { createMemoryHistory, Location, MemoryHistory } from 'history';
import merge from 'lodash.merge';
import React, { ReactElement } from 'react';
import { Provider } from 'react-redux';
import { Route, RouteProps, Router } from 'react-router-dom';
import { DeepPartial, Store } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { AppState } from './store';
import { AppAction, AppActions, AppActionsType } from './store/actions';
import { INITIAL_API_INFO_STATE } from './store/api-info/state';
import { INITIAL_FRIEND_INFO_STATE } from './store/friends/state';
import { INITIAL_HISTORY_STATE } from './store/history/state';
import { INITIAL_PERSISTED_STATE } from './store/persisted/state';
import { createAppStore } from './store/store';
import { INITIAL_USER_EVENT_STATE } from './store/user-events/state';
import { INITIAL_USER_STATE } from './store/user/state';
import { INITIAL_VIEW_STATE } from './store/view/state';
import { INITIAL_WORLD_STATE } from './store/worlds/state';

function Wrapper(
  props: React.PropsWithChildren<{
    store: Store;
    history: MemoryHistory;
    route: Pick<RouteProps, 'path'> & Partial<RouteProps>;
  }>,
): ReactElement {
  return (
    <Router history={props.history}>
      <Provider store={props.store}>
        <Route {...props.route}>{props.children}</Route>
      </Provider>
    </Router>
  );
}

export const INITIAL_TEST_APP_STATE: AppState = {
  persisted: INITIAL_PERSISTED_STATE,
  user: INITIAL_USER_STATE,
  view: INITIAL_VIEW_STATE,
  apiInfo: INITIAL_API_INFO_STATE,
  friends: INITIAL_FRIEND_INFO_STATE,
  userEvents: INITIAL_USER_EVENT_STATE,
  worlds: INITIAL_WORLD_STATE,
  history: INITIAL_HISTORY_STATE,
};

export function extendInitialState(state: DeepPartial<AppState>): AppState {
  return merge({}, INITIAL_TEST_APP_STATE, state);
}

export function createTestStore(state?: DeepPartial<AppState>): TestStore {
  return createAppStore(extendInitialState(state || {})) as TestStore;
}

export type LocationOverrides<S = Record<string, unknown>> =
  | DeepPartial<Omit<Location, 'state'>>
  | (DeepPartial<Omit<Location, 'state'>> & { state: S });

interface CustomRenderSettings<S> {
  options?: Omit<RenderOptions, 'wrapper'>;
  state?: DeepPartial<AppState>;
  location?: LocationOverrides<S>;
  route?: Partial<RouteProps>;
}

type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends (...args: never) => never ? K : never }[keyof T];
type SpyObject<T, F extends FunctionPropertyNames<T>> = Pick<T, Exclude<keyof T, F>> & {
  [P in F]: T[P] &
    jest.SpyInstance<
      ReturnType<T[P] extends (...args: never) => never ? T[P] : never>,
      Parameters<T[P] extends (...args: never) => never ? T[P] : never>
    >;
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export type HistorySpy = SpyObject<MemoryHistory<Record<string, unknown>>, 'push' | 'replace' | 'goBack'>;

export type TestStore = Store<AppState, AppActions> & {
  dispatch: ThunkDispatch<AppState, undefined, AppAction<AppActionsType>>;
};

export interface CustomRenderResult {
  result: RenderResult;
  store: TestStore;
  history: HistorySpy;
}

const initialLocation: Location<Record<string, unknown>> = {
  pathname: '/',
  search: '',
  state: {},
  hash: 'x122c123',
  key: '123',
};

// export function withAuthentication(state?: DeepPartial<AppState>): DeepPartial<AppState> {
//   return merge({}, state || {}, {
//     apiInfo: { clientApiKey: 'testApiKey' },
//     cookies: [{ key: 'auth', value: 'testAuthToken', url: 'testUrl' }],
//     userInfo: USER_INFO_FETCH_RESULT as AuthenticatedUserInfo,
//   });
// }

export function spyObject<T, F extends FunctionPropertyNames<T>>(object: T, functions: F[]): SpyObject<T, F> {
  Object.keys(object).forEach((prop) => {
    const field = prop as F;
    if (functions.includes(field)) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      //@ts-ignore
      jest.spyOn(object, field);
    }
  });

  return object as unknown as SpyObject<T, F>;
}

export function createTestHistory<S>(location?: LocationOverrides<S>): HistorySpy {
  const $location = merge({}, initialLocation, location) as Location<S>;
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const history = spyObject(createMemoryHistory(), ['push', 'replace', 'goBack']);
  history.replace($location); // move into test code?
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return history;
}

const defaultRoute: Pick<RouteProps, 'path'> = {
  path: '/',
};

function customRender<S = null>(ui: React.ReactElement, settings?: CustomRenderSettings<S>): CustomRenderResult {
  const store = createTestStore(settings?.state);
  const location = merge({}, initialLocation, settings?.location) as Location<S>;
  const history = createTestHistory<S>(location);

  const route = merge({}, defaultRoute, settings?.route);
  return {
    result: render(
      <Wrapper history={history} store={store} route={route}>
        {ui}
      </Wrapper>,
      settings?.options,
    ),
    store: store,
    history: history,
  };
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
