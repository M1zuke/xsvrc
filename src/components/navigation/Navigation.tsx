import {
  Home,
  List,
  Notifications,
  PeopleAlt,
  Person,
  PowerSettingsNew,
  Search,
  Settings,
  Shield,
} from '@mui/icons-material';
import React, { ReactElement, useCallback, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { useHistory, useLocation } from 'react-router-dom';
import { isLoaded } from '../../api/prepare';
import { useApi } from '../../api/use-api';
import { routes } from '../../common/routes';
import { selectFriendInfo } from '../../store/friends/selectors';
import { selectFavorites, selectNotifications, selectUserInfo } from '../../store/user/selectors';
import { Button } from '../button/Button';
import { Content } from '../content/Content';
import styles from './Navigation.module.scss';

export function Navigation(): ReactElement {
  const userInfo = useSelector(selectUserInfo);
  const notifications = useSelector(selectNotifications);
  const friendInfo = useSelector(selectFriendInfo);
  const favorites = useSelector(selectFavorites);
  const history = useHistory();
  const { pathname } = useLocation();
  const { logout } = useApi();

  const userId = useMemo(() => (isLoaded(userInfo) ? userInfo.id : 'none'), [userInfo]);

  const navigateTo = useCallback(
    (url: string) => {
      history.push(url);
    },
    [history],
  );
  const doLogout = useCallback(() => logout().finally(), [logout]);
  const notificationsBadge = useMemo(() => (isLoaded(notifications) ? notifications.length : 0), [notifications]);
  const isActiveRoute = useCallback((route: string) => pathname === route, [pathname]);

  if (!isLoaded(friendInfo) || !isLoaded(userInfo) || !isLoaded(notifications) || !isLoaded(favorites)) {
    return <div />;
  }

  return (
    <Content className={styles.Component}>
      <div className={styles.Navigation}>
        <Button
          aria-label="navigate to home"
          active={isActiveRoute(routes.home.path)}
          onClick={() => navigateTo(routes.home.path)}
          icon
        >
          <Home />
        </Button>
        <Button
          aria-label="navigate to friends-profile"
          active={isActiveRoute(`${routes.friendsProfile.path}/${userId}`)}
          onClick={() => navigateTo(`${routes.friendsProfile.path}/${userId}`)}
          icon
        >
          <Person />
        </Button>
        <Button
          aria-label="navigate to friends"
          active={isActiveRoute(routes.friends.path)}
          onClick={() => navigateTo(routes.friends.path)}
          icon
        >
          <PeopleAlt />
        </Button>
        <Button
          aria-label="navigate to moderations"
          active={isActiveRoute(routes.moderation.path)}
          onClick={() => navigateTo(routes.moderation.path)}
          icon
        >
          <Shield />
        </Button>
        <Button
          aria-label="navigate to notifications"
          active={isActiveRoute(routes.notifications.path)}
          onClick={() => navigateTo(routes.notifications.path)}
          badge={notificationsBadge}
          icon
        >
          <Notifications />
        </Button>
        <Button
          aria-label="navigate to event-list"
          active={isActiveRoute(routes.eventList.path)}
          onClick={() => navigateTo(routes.eventList.path)}
          icon
        >
          <List />
        </Button>
        <Button
          aria-label="navigate to user search"
          active={isActiveRoute(routes.userSearch.path)}
          onClick={() => navigateTo(routes.userSearch.path)}
          icon
          disabled
        >
          <Search />
        </Button>
      </div>
      <div className={styles.Settings}>
        <Button
          aria-label="navigate to settings"
          active={isActiveRoute(routes.settings.path)}
          onClick={() => navigateTo(routes.settings.path)}
          icon
        >
          <Settings />
        </Button>
        <Button aria-label="navigate to settings" onClick={() => doLogout()} icon>
          <PowerSettingsNew className={styles.LogoutButton} />
        </Button>
      </div>
    </Content>
  );
}
