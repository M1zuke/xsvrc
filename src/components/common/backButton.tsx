import { faAngleLeft } from '@fortawesome/free-solid-svg-icons/faAngleLeft';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React, { FC, useCallback } from 'react';
import { useHistory } from 'react-router';

import './backButton.scss';

export const BackButton: FC = () => {

  const history = useHistory();
  const goBack = useCallback(() => {
    history.goBack();
  }, [history]);

  return (
    <div className="back-btn-component" onClick={goBack}>
      <FontAwesomeIcon icon={faAngleLeft} size="3x" />
    </div>
  );
};
