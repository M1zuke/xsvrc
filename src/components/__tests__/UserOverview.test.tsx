import userEvent from '@testing-library/user-event';
import pretty from 'pretty';
import React from 'react';
import { routes } from '../../common/routes';
import { render, withAuthentication } from '../../test-utils';
import { Navigation } from '../navigation/Navigation';

describe('<Navigation />', () => {
  it('renders empty on empty storage', () => {
    const customResult = render(<Navigation />);
    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });

  it('renders correctly with userInfo', () => {
    const customResult = render(<Navigation />, { state: withAuthentication() });
    expect(pretty(customResult.result.container.innerHTML)).toMatchSnapshot();
  });

  it('navigates to friends', () => {
    const customResult = render(<Navigation />, { state: withAuthentication() });
    userEvent.click(customResult.result.getByLabelText(/navigate to friends/i));
    expect(customResult.history.location.pathname).toEqual(routes.friends.path);
  });
});
