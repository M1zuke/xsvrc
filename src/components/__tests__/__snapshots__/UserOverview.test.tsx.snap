// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserOverview /> renders correctly with userInfo 1`] = `
"<div class=\\"Component\\">
  <div style=\\"background-image: url(testUrl);\\" class=\\"UserImage\\">
    <div class=\\"UserName\\">TestUser</div>
  </div>
  <div class=\\"UserName\\">TestUser</div>
  <div class=\\"UserName\\"><button aria-label=\\"navigate to friends\\" class=\\"Component\\"><span class=\\"StyleElement\\">Friends</span></button></div>
</div>"
`;

exports[`<UserOverview /> renders empty on empty storage 1`] = `""`;
